File: c/v6d.c
#include <stdint.h>

union Pixel
{
    uint8_t c[4]; // four channels: red, green, blue, alpha
    uint32_t v;   // full pixel value as a 32-bit integer
};

void v6_darken(union Pixel* first, union Pixel* last, int darkness)
{
  uint32_t lightness = 256 - darkness;
  for (; first < last; ++first) {
      uint32_t p = first->v;
      uint32_t c0 = ((p & 0xff) * lightness / 256) & 0xff;
      uint32_t c1 = ((p & 0xff00) * lightness / 256) & 0xff00;
      uint32_t c2 = ((p & 0xff0000) * lightness / 256) & 0xff0000;
      uint32_t c3 = p & 0xff000000;
      first->v = c0 | c1 | c2 | c3;
  }
}

File: rust/src/v6d.rs
use crate::Pixel;

pub fn darken(buf: &mut [Pixel], darkness: i32) {
    let (prefix, buf, suffix) = unsafe { buf.align_to_mut::<u32>() };
    assert_eq!(prefix.len(), 0);
    assert_eq!(suffix.len(), 0);

    let lightness = (256 - darkness) as u32;
    for p in buf {
        let v = *p;
        let r = ((v & 0xff) * lightness / 256) & 0xff;
        let g = ((v & 0xff_00) * lightness / 256) & 0xff_00;
        let b = ((v & 0xff_00_00) * lightness / 256) & 0xff_00_00;
        let a = v & 0xff_00_00_00;
        *p = r | g | b | a;
    }
}

#[bench]
fn bench(b: &mut test::Bencher) {
    let mut buf = crate::random_image(test::black_box(1_000_000));
    let darkness = test::black_box(8);
    b.iter(|| darken(&mut buf, darkness));
}
